tests/factories/lti_user.py

```diff
@@ -0,0 +1,10 @@
+import factory
+
+from lms import values
+
+LTIUser = factory.make_factory(  # pylint:disable=invalid-name
+    values.LTIUser,
+    user_id=factory.Faker("hexify", text="^" * 40),
+    oauth_consumer_key=factory.Faker("hexify", text="Hypothesis" + "^" * 32),
+    roles=factory.Faker("random_element", elements=["Learner", "Instructor"]),
+)
```

Comment:
For this simple factory I'm using Factory Boy's [lightweight factory declaration](https://factoryboy.readthedocs.io/en/latest/reference.html#lightweight-factory-declaration), which unfortunately requires a `# pylint:disable=invalid-name`.

The longer-form way, which you need to use if your factory requires advanced Factory Boy features, would be:

```python
class LTIUser(factory.Factory):
    class Meta:
        model = values.LTIUser

    user_id = factory.Faker("hexify", text="^" * 40)
    oauth_consumer_key = factory.Faker("hexify", text="Hypothesis" + "^" * 32)
    roles = factory.Faker("random_element", elements=["Learner", "Instructor"])
```

This also has an unfortunate PyLint shortcoming: PyLint understandably thinks that `factories.LTIUser()` returns an object of class `factories.LTIUser`. When you then try to access `values.LTIUser` fields on that object (e.g. `lti_user.is_learner`) PyLint complains that the field doesn't exist.

Reply:
The Factory Boy docs tend to name the factory classes with a `...Factory` suffix like `UserFactory` etc. You'd then do:

```python
from tests.factories import UserFactory

test_user = UserFactory(...)
```

I've followed h's approach of just calling the factory `User`:

```python
from tests import factories

test_user = factories.User(...)
```

This works well when you **have** to access factories via a `factories` fixture as you do for DB factories in h and may in future need to for future DB factories in LMS. You can just write `factories.User()` not `factories.UserFactory()`

Reply:
The `Faker` stuff deserves some explanation:

* Factory Boy integrates a library called [Faker](https://faker.readthedocs.io/) for randomly generating realistic looking fake values.

* Faker comes with a set of "providers": functions for returning different types of fake values. See [Faker's built-in providers](https://faker.readthedocs.io/en/latest/providers.html) and [Faker's community providers](https://faker.readthedocs.io/en/latest/communityproviders.html)

* The way you call Faker via Factory Boy is `factory.Faker("provider", **kwargs)` where `"provider"` is the name of the Faker provider method as a string and `**kwargs` are the kwargs for Factory Boy to pass to the provider method (see the docs for each individual provider method for what kwargs it supports).

  So `factory.Faker("hexify", text="Hypothesis" + "^" * 32)` calls Faker's [`hexify()`](https://faker.readthedocs.io/en/latest/providers/baseprovider.html#faker.providers.BaseProvider.hexify) with `text="Hypothesis" + "^" * 32`.

* `hexify()` is one of a few flexible Faker methods for generating random strings. It replaces every `"^"` in the given string with a random hex character. So passing it `text="Hypothesis" + "^" * 32` causes it go generate strings like `"Hypothesis<32_RANDOM_HEX_CHARS>"`, which look like the real consumer keys that the LMS app generates.

  There's also `lexify()` which replaces `"?"`'s with random letters, `numerify()` which replaces `"#"`'s, `"%"`'s, `"!"`'s and `"@"`'s with different kinds of digits, and `bothify()` which combines `lexify()` and `numerify()`.

* Factory Boy passes its fixed random seed to Faker and everything else that Factory Boy uses that requires a randomness source, so they all use the same predictable randomness.

* You can also add your own providers to Faker. See Faker's [docs on writing your own provider](https://faker.readthedocs.io/en/latest/index.html#how-to-create-a-provider) and Factory Boy's [docs on adding a provider to Factory Boy's Faker instance](https://factoryboy.readthedocs.io/en/latest/reference.html#factory.Faker.add_provider).

  It seems unlikely you'll need to add a provider though. Faker comes with a pretty good set. And if there's an existing provider that's close to what you want but not quite right you can use various Factory Boy tricks to adjust what Faker returns, the easiest of which is probably Factory Boy's [`_adjust_kwargs()` hook](https://factoryboy.readthedocs.io/en/latest/reference.html#factory.Factory._adjust_kwargs)

See Factory Boy's [docs about its Faker integration](https://factoryboy.readthedocs.io/en/latest/reference.html#faker)

Reply:
The factory of course generates realistic-looking random `LTIUser`'s. They look more realistic than what all the tests are using currently: they have a realistic, randomly generated `user_id` and `oauth_consumer_key` etc:

```terminal
$ .tox/tests/bin/python
Python 3.6.9 (default, Jan  7 2020, 12:11:53) 
[GCC 9.2.1 20191008] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from tests import factories
>>> factories.LTIUser()
LTIUser(user_id='52d26561232aa962634d245a1fec2804f00a3804', oauth_consumer_key='Hypothesis1371da4380dc333f2075fcbcfc7680bb', roles='Instructor')
>>> factories.LTIUser()
LTIUser(user_id='2b402c900d752a4867223f0e8651e5d3a7784e98', oauth_consumer_key='Hypothesis5321d75bd9f8ee356b890140262b6ea5', roles='Learner')
>>> factories.LTIUser()
LTIUser(user_id='8fad39e652ded224dcf5d70765ca71988f4a98b5', oauth_consumer_key='Hypothesis53dc1cb8beb0c86a1ea05682e377e26f', roles='Instructor')
>>> 
```

Reply:
`bothify()` is a spectacular name.